File: src/lib/stores/schema.ts
import { writable } from 'svelte/store';

export interface TestConfigSchema {
  test_info: {
    id: string;
    display_name: string;
    description: string;
    category: string;
    icon: string;
    color: string;
  };
  contextual_description: string;
  compatibility: 'Compatible' | 'Incompatible' | 'Conditional';
  requirements_met: boolean;
  compatibility_reason?: string;
  fields: ConfigField[];
  warnings: ValidationMessage[];
  errors: ValidationMessage[];
}

export interface ConfigField {
  id: string;
  label: string;
  field_type: {
    base_type: string;
    constraints: Record<string, any>;
    options?: SelectOption[];
  };
  required: boolean;
  default_value?: any;
  help_text?: string;
  placeholder?: string;
  advanced: boolean;
}

export interface SelectOption {
  value: string;
  label: string;
  description?: string;
  disabled: boolean;
}

export interface ValidationMessage {
  message: string;
  field_id?: string;
  severity: 'Error' | 'Warning' | 'Info';
}

export interface NodeContextForAPI {
  node_id?: string;
  node_type: string;
  capabilities: string[];
  target: any; // Just pass through the existing NodeTarget
  assigned_tests: string[];
}

export async function fetchTestSchema(testType: string, nodeContext: NodeContextForAPI): Promise<TestConfigSchema> {
  const response = await fetch('/api/tests/schema', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      test_type: testType,
      node_context: nodeContext,
    }),
  });
  
  if (!response.ok) {
    throw new Error(`Failed to fetch schema: ${response.statusText}`);
  }
  
  const data = await response.json();
  if (!data.success) {
    throw new Error(data.error || 'Failed to fetch schema');
  }
  
  return data.data;
}

File: src/server/tests/types/schema.rs
use serde::{Deserialize, Serialize};
use crate::server::nodes::types::capabilities::NodeCapability;
use crate::server::nodes::types::types::NodeType;

use crate::server::shared::types::metadata::TypeMetadata;
use crate::server::{
    nodes::types::{targets::NodeTarget}
};

#[derive(Serialize, Debug, Clone)]
pub struct TestConfigSchema {
    pub test_info: TypeMetadata,
    pub contextual_description: String,
    pub compatibility: CompatibilityStatus,
    pub requirements_met: bool,
    pub compatibility_reason: Option<String>,
    pub fields: Vec<ConfigField>,
    pub warnings: Vec<ValidationMessage>,
    pub errors: Vec<ValidationMessage>,
}

#[derive(Serialize, Debug, Clone, PartialEq)]
pub enum CompatibilityStatus {
    Compatible,
    Incompatible,
    Conditional,
}

#[derive(Serialize, Debug, Clone)]
pub struct ConfigField {
    pub id: String,
    pub label: String,
    pub field_type: FieldType,
    pub required: bool,
    pub default_value: Option<serde_json::Value>,
    pub help_text: Option<String>,
    pub placeholder: Option<String>,
    pub advanced: bool,
}

#[derive(Serialize, Debug, Clone)]
pub struct FieldType {
    pub base_type: String,
    pub constraints: serde_json::Value,
    pub options: Option<Vec<SelectOption>>,
}

#[derive(Serialize, Debug, Clone)]
pub struct SelectOption {
    pub value: String,
    pub label: String,
    pub description: Option<String>,
    pub disabled: bool,
}

#[derive(Serialize, Debug, Clone)]
pub struct ValidationMessage {
    pub message: String,
    pub field_id: Option<String>,
    pub severity: MessageSeverity,
}

#[derive(Serialize, Debug, Clone)]
pub enum MessageSeverity {
    Error,
    Warning,
    Info,
}

#[derive(Debug, Clone, Deserialize)]
pub struct NodeContext {
    pub node_id: Option<String>,
    pub node_type: NodeType,
    pub capabilities: Vec<NodeCapability>,
    pub target: NodeTarget,
}